errors.go

```diff
@@ -118,86 +126,138 @@ func (p *Error) LogMetadata() map[string]string {
 // New creates a new error for you. Use this if you want to pass along a custom error code.
 // Otherwise use the handy shorthand factories below
 func New(code string, message string, params map[string]string) *Error {
-	return errorFactory(code, message, params)
+	return errorFactory(nil, code, message, params)
+}
+
+// Propagate takes any error interface and wraps it into an Error.
+// This is useful because an Error contains lots of useful goodies, like the stacktrace of the error.
+// NOTE: If `cause` is already an `Error`, it will merge the message and params with the underlying Error
+func Propagate(cause error, msg string, params map[string]string) error {
+	return PropagateWithCode(cause, ErrInternalService, msg, params)
+}
+
+// PropagateWithCode wraps an error with a custom error code.
+// NOTE: If `cause` is already an `Error`, it will merge the message and params with the underlying Error
+func PropagateWithCode(cause error, code string, msg string, params map[string]string) error {
+	if cause == nil {
+		return nil
+	}
+
+	switch cause := cause.(type) {
+	case *Error:
+		errString := cause.Message
+		if msg != "" {
+			errString = fmt.Sprintf("%s: %s", msg, cause.Message)
+		}
+
+		return &Error{
+			Code:        cause.Code,
+			Message:     errString,
+			Params:      merge(cause.Params, params),
+			StackFrames: cause.StackFrames,
```

Comment:
We should look at whether this is the right thing to do now that we're able to properly wrap and unwrap errors. It effectively means that the `StackFrames` is set the the "deepest" terror.